### PR TITLE
Gracefully handle analyzer diagnostics with location outside the current compilation

### LINKED ISSUE #7173: Visual Studio 2015 Update 1 crashing with Code Cracker
Originally reported by @Cybermaxs on https://github.com/code-cracker/code-cracker/issues/610#issuecomment-160966346.

Here is the stack trace from his event logs:

````
Message : System.ArgumentException: SyntaxTree ' {{CsCodeRemovedForClarity}}'
not found to remove
Parameter name: syntaxTree
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(String id, Location location, SuppressMessageInfo& info)
   at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Diagnostic diagnostic, Compilation compilation, ISymbol symbolOpt)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSource(ImmutableArray`1 diagnostics, Compilation compilation)
   at Microsoft.CodeAnalysis.Diagnostics.AnalysisResult.StoreAnalysisResult(AnalysisScope analysisScope, AnalyzerDriver driver, Compilation compilation)
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<ComputeAnalyzerDiagnosticsCoreAsync>d__54.MoveNext()
Pile :
   à System.Environment.FailFast(System.String, System.Exception)
   à Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   à Microsoft.CodeAnalysis.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   à Microsoft.CodeAnalysis.FatalError.ReportUnlessCanceled(System.Exception)
   à Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__54.MoveNext()
   à Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(Microsoft.CodeAnalysis.SyntaxTree, Boolean)
   à Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetSemanticModel(Microsoft.CodeAnalysis.SyntaxTree, Boolean)
   à Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(System.String, Microsoft.CodeAnalysis.Location, Microsoft.CodeAnalysis.Diagnostics.SuppressMessageInfo ByRef)
   à Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Microsoft.CodeAnalysis.Diagnostic, Microsoft.CodeAnalysis.Compilation, Microsoft.CodeAnalysis.ISymbol)
   à Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.FilterDiagnosticsSuppressedInSource(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.Diagnostic>, Microsoft.CodeAnalysis.Compilation)
   à Microsoft.CodeAnalysis.Diagnostics.AnalysisResult.StoreAnalysisResult(Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver, Microsoft.CodeAnalysis.Compilation)
   à Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__54.MoveNext()
   à System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<ComputeAnalyzerDiagnosticsCoreAsync>d__54, Microsoft.CodeAnalysis, Version=1.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ComputeAnalyzerDiagnosticsCoreAsync>d__54 ByRef)
   à Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsCoreAsync(Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver, Microsoft.CodeAnalysis.Diagnostics.AsyncQueue`1<Microsoft.CodeAnalysis.Diagnostics.CompilationEvent>, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, System.Threading.CancellationToken)
   à Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass51_2+<<ComputeAnalyzerDiagnosticsAsync>b__1>d.MoveNext()
   à System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass51_2+<<ComputeAnalyzerDiagnosticsAsync>b__1>d, Microsoft.CodeAnalysis, Version=1.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<<ComputeAnalyzerDiagnosticsAsync>b__1>d ByRef)
   à Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers+<>c__DisplayClass51_2.<ComputeAnalyzerDiagnosticsAsync>b__1()
   à System.Threading.Tasks.Task`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].InnerInvoke()
   à System.Threading.Tasks.Task.Execute()
   à System.Threading.Tasks.Task.ExecutionContextCallback(System.Object)
   à System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   à System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   à System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
   à System.Threading.Tasks.Task.ExecuteEntry(Boolean)
   à System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   à System.Threading.ThreadPoolWorkQueue.Dispatch()
   à System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
````

### PR DESCRIPTION
**Pull request by mavasani**
Code cracker recently added an analyzer that accidentally reports diagnostics with a location in a syntax tree outside of the current compilation. We were not handling this case gracefully and causing VS to crash. Code cracker fixed their analyzer and released an updated package. This change guards against such cases by ignoring such diagnostics and reporting an analyzer exception diagnostic about invalid diagnostic.

Fixes #7173

### CODE REVIEW

**Comment by mavasani** [comment]
@JohnHamby @gafter @srivatsn @heejaechang

**Comment by mavasani** [comment]
Ping.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 1]
```diff
@@ -751,10 +757,11 @@ public struct OperationBlockAnalysisContext
         /// </summary>
         public CancellationToken CancellationToken => _cancellationToken;
 
-        public OperationBlockAnalysisContext(ImmutableArray<IOperation> operationBlocks, ISymbol owningSymbol, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
+        public OperationBlockAnalysisContext(ImmutableArray<IOperation> operationBlocks, ISymbol owningSymbol, SemanticModel semanticModel, AnalyzerOptions options, Action<Diagnostic> reportDiagnostic, Func<Diagnostic, bool> isSupportedDiagnostic, CancellationToken cancellationToken)
```
Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Why is this ctor public?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs [hunk 2]
```diff
@@ -898,7 +918,12 @@ public struct OperationAnalysisContext
         /// <see cref="IOperation"/> that is the subject of the analysis.
         /// </summary>
         public IOperation Operation { get { return _operation; } }
-        
+
+        /// <summary>
+        /// <see cref="CodeAnalysis.SemanticModel"/> that can provide semantic information about the <see cref="IOperation.Syntax"/> associated with the <see cref="Operation"/>.
+        /// </summary>
+        public SemanticModel SemanticModel { get { return _semanticModel; } }
```
Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Not sure why public API changes are included in this PR.

**Comment by gafter** [comment]
Looks good, except for the public API changes. Not sure why they are mixed up in this PR.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: All the context type constructors were made public prior to 1.0.0 release to enable 3rd party analyzer authors to write their own test frameworks to run analyzers. Apart from that, there should be no public consumers that want to instantiate context types as the analyzer driver does it for them.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Validating that a diagnostic reported by the analyzer has location within a compilation requires the analysis context to have the compilation being analyzed. All the context types that shipped prior to recently added Operation analysis contexts take either the semantic model/compilation parameter in the constructor and expose the same it publically on the context, but the Operation related contexts don't hold onto or expose either of these. This change makes the public API shape of Operation context types consistent with other context types by exposing the Semantic model for the tree corresponding to operation being analyzer and also passing down the owning compilation of the model to the code validating diagnostic location - I can rip out the public API changes by making these internal if you insist. @JohnHamby what is your take on this?

Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: In theory the semantic model is not useful to an operation action. The main reason for making the syntax node available is to be able to get location information. Now that we're thinking about mapping IL to IOperation, replacing the syntax node property from IOperation with a location property might be reasonable. If we do that, then a semantic model would be completely useless to an author.

So, I don't think the semantic model property should be public here.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: I didn't want any context constructors to be public, but we decided to make them all public for testability.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs]: Ok, I will switch the properties to be internal then.

**Comment by mavasani** [comment]
@gafter @JohnHamby Updated the PR reverting all the public API changes.

**Pull request merged**